iceoryx_posh/source/roudi/process_manager.cpp

```diff
@@ -416,6 +416,19 @@ void ProcessManager::addSubscriberForProcess(const RuntimeName_t& name,
 {
     findProcess(name)
         .and_then([&](auto& process) {
+            auto segmentInfo = m_segmentManager->getSegmentInformationWithReadAccessForUser(process->getUser());
```

Comment:
I may be wrong but I think this is the wrong approach. Here is why.
1. Let's assume publisher A is on shared memory segment A to which the subscriber B has no read access
2. But subscriber B has read access to shared memory segment B

When subscriber B would like to receive data it has to read from segment A.

With this algorithm you are just looking for any readable segment and not if segment A of the matching publisher is readable. Now the following happens.
1. Subscriber B is created since segment B is readable
2. Publisher A sends a sample to Subscriber B
3. Subscriber B is accessing the memory to which the sample is pointing and SEGFAULT.

Reply:
@elfenpiff You're not wrong here. It is exactly as you described. IMHO the long-term solution is to have the groups stored in either the `ServiceDescription` or `popo::Publisher` and `popo::Subscriber` to be able to see if they match in `isCompatible{PubSub, ClientServer}`. However, this would be a major refactoring and is therefore out of scope for `v2.0`. This is something we should tackle together with the `ServiceDescription` refactoring in #1141.

For `v2.0` instead of silently failing with a segfault, which can lead to nasty debugging sessions for several days, I would like to add the proposed patches to fail early if the process has no access to any shared memory segment. This makes the world a little bit better and follows our iterative improvement process which we aim to establish for iceoryx.

@elBoberido @MatthiasKillat @elBoberido Would you prefer a segfault or a clear and concise error message for this case?

Reply:
@mossmaurice But your patch does not solve this - only the rare corner case that an application has neither read nor write access to any segment. But in this case we should ask ourselfs why can this application register itself at RouDi at all?

So for this corner case - app has no read/write access to anything - I would propose that RouDi denies the registration of the app with a NACK response to the initial register request.

But with your solution we would still end up in a segfault when the app has read access to something so it does not solves the problem.

I would not merge this into `v2.0` and move this whole issue to `v3.0` especially since time (review time & coding time) is precious this week.

Final note: `Unit tests have been written for new behavior` something I have to check as a reviewer and I cannot cross out since the (integration) tests for this are missing.

Reply:
>But with your solution we would still end up in a segfault when the app has read access to something so it does not solves the problem.

@elfenpiff Well, it's exactly this 'corner case' that we describe in the `ice_access_control` example. Imagine someone 
would copy and paste it and then run into the segfault. I completely agree that a `NACK` on registration makes more sense in the long run. I've sketched a hierarchical management of the entities in shared memory on #620. My proposal is to implement RAII e.g. if a node is destroyed also publisher, clients, etc. inside are destroyed.

> Final note: Unit tests have been written for new behavior something I have to check as a reviewer and I cannot cross out since the (integration) tests for this are missing.

Just yesterday a PR got merged with your approval without proper integration tests :clown_face: The [explanation](https://github.com/eclipse-iceoryx/iceoryx/pull/1137#discussion_r812065503) of @elBoberido makes sense to me.

Reply:
@mossmaurice I think the proper way to avoid this issue with the SEGFAULT the information about the segment which the publisher is writing to needs to be propagated to the subscriber when they initially establish connection. If this segment is in the readable list of the app, then they get connected. Otherwise the publisher will not add the corresponding susbcriber to its list of queues. This can be part of some publisher settings , essentially a private(i.e. not directly user configurable) part of the options which are considered during the matching process. We can also consider later the option to configure a segment affinity for publishers.

Later on another publisher which writes to a segment we can read from may become available, so we should not deny creation of the subscriber (and even a warning is questionable, it will just get no data in the meantime and not be connected).

Currently we fix only the SEGFAULT if we cannot read from any segment. This is better than before but still does not solve the main issue. Whether this is valuable enough for us right now I cannot say, but it is at least not worse than before.